### ISSUE TITLE
[Enhancement] Introducing External JMX Access

### ISSUE THREAD

**Issue by julianGoh17** [strimzi/strimzi-kafka-operator#2209]
**Is your feature request related to a problem? Please describe.**
Currently, Strimzi has a JMX exporter that runs along side the kafka broker to scrape metrics from the broker and push it to Prometheus. There is currently no other way of exposing the JMX port within the cluster to get JMX data from the Kafka brokers without Prometheus. 

We also want to make sure that the port exposed is password and username protected as a malicious user could introduce an external JMX agent which could connect to the JMX port to disrupt the Kafka broker by invoking Mbean Operations on the Kafka Broker.

The reason why we need an externally accessible port can be seen through the following example: Imagine that we have a splunk enterprise server that is accessible at `www.splunkRemoteHost.com`. We only care about certain statistics such as the BrokerTopicMetrics and specifically want to get the mean rate. In Strimzi's current state, we would need to deploy a prometheus pod to read that data and then deploy another pod which will get the data from prometheus and push it to `www.splunkRemoteHost.com`. However, with the method proposed below after passing in some yaml, the jmxtrans pod will automatically be configured to read that data from this endpoint and push it to `www.splunkRemoteHost.com`. Furthermore, we can make sure to secure the JMX port with a username and password to prevent a malicious user from deploying a JMX agent to perform mbean Operations on the kafka broker to disrupt the broker.

**Describe the solution you'd like**
Add an intermediate JMX pod with https://github.com/jmxtrans/jmxtrans as the base. By introducing the JMX Trans Pod, we can use it read specific queries on various Mbean Attributes of the Kafka Broker and pushes it to various sinks (hosts of information). What this pod would allow us to do is to introduce a password protected JMX port on the broker which we can connect to with a service to the JMX port and push data out to a variety of sinks in various data formats (such as graphite). I would like to create a new watcher for this port and some new configuration in the CRD to allow you to configure this data. 

In this yaml file, jmxtrans will be pushing out data about BrokerTopicMetrics's count and meant rate from JMX port that is not secured with a username or password to standard output. A simplified yaml configuration can be seen below. : 
```yaml
metrics: 
  hasAuthentication: false
  jmxTrans:
    outputDefinitions:
      - outputType: "com.googlecode.jmxtrans.model.output.StandardOutput"
        name: standardOutput
    queries:  
      - targetAttribute: "kafka.server:type=BrokerTopicMetrics,name=*"
        attr: ["Count", "MeanRate"]
        outputs:
         - standardOutput
    
```

A more complex yaml file can be seen below. We have defined an additional logstash container which runs on localhost 9999  which will configure a JMX trans pod to push the following data from a JMX port that is username and password protected:
- **logstash:** Getting graphite data about the BrokerTopicMetrics's Count and MeanRate and DelayedOperationPurgatory's Fetch and Topics
- **StandardOutput:** data about the BrokerTopicMetrics's Count and MeanRate 

```yaml
metrics: 
  hasAuthentication: true
  jmxTrans:
    outputDefinitions:
      - outputType: "com.googlecode.jmxtrans.model.output.GraphiteWriterFactory"
        port: "9999" 
        outputHost: "localhost" 
        flushDelay: 5 
        name: logstash
      - outputType: "com.googlecode.jmxtrans.model.output.StandardOutput"
        name: standardOutput
    queries:  
      - targetAttribute: "kafka.server:type=BrokerTopicMetrics,name=*"
        attr: ["Count", "MeanRate"]
        outputs:
         - splunk
         - standardOutput
      - targetAttribute: "kafka.server:type=DelayedOperationPurgatory,name=*"
        attr: ["Fetch", "Topic"]
        outputs:
         - splunk
    additionalContainers:
          - name: logstash
             env:
                 LISTENING_PORT: 9999
```

Ideally you would also be able to specify an additional container you want to put in the JMX trans pod to push data to if you don't have a remote server.

After specifying the configuration for the JMX Trans pod, it will automatically create the queries and push it to the remote hosts for each broker. This ensures that if a user were to scale up/scale down the number of brokers, then the JMX trans pod will automatically reconfigure itself to match the number brokers specified.

Once the user has specified that `hasAuthentication: true` then the JMX port will automatically generate a username and password to secure the port (on start up of the kafka broker) and then store the credentials in a secret. That way a malicious user can't read the jmxtrans config to get the username and password.

The following diagram is what the proposed plan would look like:
<img width="926" alt="Screen Shot 2019-11-28 at 11 16 44" src="https://user-images.githubusercontent.com/44057265/69802059-9cd07b80-11d0-11ea-8438-8c2407afcee3.png">

**Describe alternatives you've considered**
**Exposing JMX Directly Externally** There is no way of exposing the JMX port directly externally through the cluster with Routes because you will need to set the RMI port to 443 or 80, which are protected ports that one can't expose on the Kafka Broker. You can however, do this with NodePorts but this doesn't work with Openshift 4.2

**Comment by tomaley** [comment]
To clarify on the point
> You can however, do this with NodePorts but this doesn't work with Openshift 4.2

OpenShift 4.x installs on AWS - The first platform that Installer Provisioned Infrastructure (IPI) installers supported, deploys a load-balancer in front of the OpenShift nodes, which are not publicly accessible. The load-balancer forwards http(80) and https(443) traffic to OpenShift Router pods allowing Routes to be used for ingress but the ports 80 and 443 cannot be assigned as RMI ports.

**Comment by julianGoh17** [comment]
Interesting point of discussion would be how we would handle the scaling up and scaling down of brokers. Jmxtrans unfortunately doesn't have dynamic reconfiguration which means that if one were to scale up or scale down the brokers the jmxtrans pod wouldn't pick it up. Hence, you won't get metrics from the newly brought up brokers.

Currently the way I have it configured is: A ConfigMap stores the config file for the jmxtrans to start up.  The config file stores a list of brokers and the queries that will be sent to the brokers. The deployment of the jmxtrans pod mounts this configmap as a volume to pick up this config file. This config is referenced on start up and once the Jmxtrans process starts, it won't pick up any changes. Hence, if we scale up the number of brokers the old config won't be valid.

I think one way to do this is to add a label into the spec of the jmxtrans deployment which refers to the revision number of the configmap. For example, when we scale up/down the brokers the controller should update the configmap and hence the configmap's revision number gets updated. The controller then writes the updated revision number to the deployment which will trigger a new jmxtrans pod to be brought up and once it is ready then the old pod goes down. Hence the flow would look like:
`User scales up replica -> ConfigMap gets updated -> ConfigMap's revision number increases -> Jmxtrans deployment spec updates the label with the new revision number -> A new jmxtrans pod comes up and once ready the old pod is torn down.`

This would also solve the problem about what happens if a user were to reconfigure the jmxtrans queries used as the same process will trigger a new pod to be brought up.

Any thoughts or suggestions on this approach? Is there an alternative approach that could be achieved with the operator?

**Comment by scholzj** [comment]
@julianGoh17 Some points to this:
* Do we do this only for Kafka? Or also for Zookeeper, Connect, etc.?
* Your examples seem to reuse the `metrics` field. It is not completely clear to me how will you implement it in the code. The `metrics` field is currently basically an opaque configuration for the Prometheus JMX exporter. It might be hard to parse it out and decide on what the user actually wants - Prometheus export or JMX Trans.
* The config map update trigger should work. But in general I think annotation should be prefered over this than label.
* Could you please describe a bit more what changes will be needed for the Kafka pods and what will be in a separate JMX Trans deployment?

**Comment by julianGoh17** [comment]
@scholzj  In response to your comment: 
1. Currently I am only doing this for Kafka. If there is a demand/use for getting metrics from the others, I think that it wouldn't be too hard to extend the set up in a later PR
2. I did notice that it was very hard to pass out what the user wanted to do so I have renamed it to be jmxMetrics (as a temporary name for now) keeping prometheus metrics as a separate field in the CRD. I am not sure what would happen if we enable both as they are currently listening on the same port. I will test this and if there are problems I will change it so that jmxMetrics will listen on a different port.
3. I have changed it to use to an annotations as I noticed that you already do that when you add the annotation for the cluster operator generation
3. All the kafka pods need is some update in the `kafka_run.sh` to add a username and password to the java start up operation. Not sure what you mean by jmxtrans in a separate deployment but what I mean is that the jmxtrans should be a deployment and not an extra container in the kafka brokers. This would allow for the ability to add additional containers to listen on localhost (like logstash) in the crd so that the deployment of the jmxtrans now has jmx and whatever extra containers the user has defined. What has been added to the jmxtrans deployment is: 
- The ability to read from the CRD 
- Creating a configmap containing the config file needed to run jmx trans
- A secret containing the username and password to connect to a secured jmx port.
- Exposing the metrics port on the kafka broker headless service.

**Comment by scholzj** [comment]
Right. So could you please share an example CR how it might look like with all the stuff enabled? I think that would help to give a better idea of how it looks. Thanks.

**Comment by julianGoh17** [comment]
The below is an example of a CR which has everything enabled:
```
kafka:
**
  jmxMetrics:
    secure: true
    outputDefinitions:
      - outputType: "com.googlecode.jmxtrans.model.output.StdOutWriter"
        name: "standardOut"
      - outputType: "com.googlecode.jmxtrans.model.output.GraphiteWriter"
        host: "http://myLogstash.com"
        port: 31028
        flushDelayInSeconds: 5
        name: "logstash"
    queries:
      - targetMBean: "kafka.server:type=BrokerTopicMetrics,name=*"
        attributes:  ["Count"]
        outputs: ["standardOut", "logstash"]
**
```
The CR above will give you the ability to define where metrics are to be sent to the remote sink (defined in the `outputDefinition`). In the `queries` section, you define the queries you want to send to the brokers and per query you pass in an array of outputDefinitions which represent where the information is being pushed to. So, you define which hosts get the information per query.

In this example, we are getting `BrokerTopicMetrics` and pushing it to standard out and a remote logstash on the host http://myLogstash.com on port 31028. The operator will also automatically generate a username and password to secure the JMX port on the brokers which is kept in a secret and later referenced in the brokers and Jmxtrans deployment.

cc: @scholzj

**Comment by ppatierno** [comment]
The configuration we have in the `jmxMetrics` field seems to be more specific to the jmxtrans project so maybe we should call it just `jmxTrans` as we do for `kafkaExporter`. I don't think we expect to support one more project as jmxtrans and having a sort of general configuration declaration which is valid for both of them.

**Comment by julianGoh17** [comment]
I think you are absolutely correct! I am currious what people think about the situation where a user just wants to open the JMX port and not deploy a JMXTrans deployment. I think that we should give the ability to open up the port via a service and add security if they wanted. 

Another thing I am concerned about is that there is already a metrics field that is specific to prometheus metrics. Maybe we want to rename the Metrics field to `prometheusMetrics`as well?

So I think that a solution to the problem could be to create a config file that looks like:
```
kafka:
**
  prometheusMetrics: 
       **
  jmxRemote:
    secure: true
    jmxTrans:
      outputDefinitions:
        - outputType: "com.googlecode.jmxtrans.model.output.StdOutWriter"
          name: "standardOut"
        - outputType: "com.googlecode.jmxtrans.model.output.GraphiteWriter"
          host: "http://myLogstash.com"
          port: 31028
          flushDelayInSeconds: 5
          name: "logstash"
      queries:
        - targetMBean: "kafka.server:type=BrokerTopicMetrics,name=*"
          attributes:  ["Count"]
          outputs: ["standardOut", "logstash"]
**
```

What do people think about this solution?

**Comment by scholzj** [comment]
Done in #2268
